redbot/cogs/mutes/mutes.py

```diff
@@ -0,0 +1,1529 @@
+            # we need to update the cache early so it's there before we receive the member_update event
+            if guild.id not in self._server_mutes:
+                self._server_mutes[guild.id] = {}
+
+            self._server_mutes[guild.id][user.id] = {
+                "author": author.id,
+                "member": user.id,
+                "until": until.timestamp() if until else None,
+            }
+            try:
+                await user.add_roles(role, reason=reason)
+                await self.config.guild(guild).muted_users.set(self._server_mutes[guild.id])
+            except discord.errors.Forbidden:
+                if guild.id in self._server_mutes and user.id in self._server_mutes[guild.id]:
+                    del self._server_mutes[guild.id][user.id]
+                ret["reason"] = _(MUTE_UNMUTE_ISSUES["permissions_issue"])
+                return ret
+            ret["success"] = True
+            return ret
+        else:
+            perms_cache = {}
+            tasks = []
+            for channel in guild.channels:
+                tasks.append(self.channel_mute_user(guild, channel, author, user, until, reason))
+            task_result = await asyncio.gather(*tasks)
+            for task in task_result:
+                if not task["success"]:
+                    ret["channels"].append((task["channel"], task["reason"]))
+                else:
+                    chan_id = task["channel"].id
+                    perms_cache[str(chan_id)] = task.get("old_overs")
+                    ret["success"] = True
+            await self.config.member(user).perms_cache.set(perms_cache)
+            return ret
+
+    async def unmute_user(
+        self,
+        guild: discord.Guild,
+        author: discord.Member,
+        user: discord.Member,
+        reason: Optional[str] = None,
+    ) -> Dict[
+        str,
+        Union[bool, Optional[str], List[Tuple[discord.abc.GuildChannel, str]], discord.Member],
+    ]:
+        """
+        Handles unmuting users
+        """
+        ret: Dict[
+            str,
+            Union[bool, Optional[str], List[Tuple[discord.abc.GuildChannel, str]], discord.Member],
+        ] = {
+            "success": False,
+            "reason": None,
+            "channels": [],
+            "user": user,
+        }
+        mute_role = await self.config.guild(guild).mute_role()
+        if mute_role:
+            if not await self.is_allowed_by_hierarchy(guild, author, user):
+                ret["reason"] = _(MUTE_UNMUTE_ISSUES["hierarchy_problem"])
+                return ret
+            role = guild.get_role(mute_role)
+            if not role:
+                ret["reason"] = _(MUTE_UNMUTE_ISSUES["role_missing"])
+                return ret
+
+            if guild.id in self._server_mutes:
+                if user.id in self._server_mutes[guild.id]:
+                    del self._server_mutes[guild.id][user.id]
+            if not guild.me.guild_permissions.manage_roles:
```

Comment:
As said in the previous review comment
> The same goes for checking whether the role we're trying to assign can be assigned by the bot by checking its position against the bot's top role.

If you want to, you can split this into 2 errors, but I won't require that.
```suggestion
            if (
                not guild.me.guild_permissions.manage_roles
                or role >= guild.me.top_role
            ):
```